// @kanaries/graphic-walker/dist/components/fieldConfigDialog/index.js
import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { observer } from 'mobx-react-lite';
import { useTranslation } from 'react-i18next';
import { Dialog, DialogContent, DialogFooter, DialogHeader, DialogTitle } from "../ui/dialog";
import { Button } from "../ui/button";
import { Input } from "../ui/input";
import { Label } from "../ui/label";
import { RadioGroup, RadioGroupItem } from "../ui/radio-group";
import { Textarea } from "../ui/textarea";
import { ScrollArea } from "../ui/scroll-area";
import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from "../ui/select";
import Spinner from "../spinner";
import Tooltip from "../tooltip";
import { useCompututaion, useVizStore } from "../../store";
import { fieldStat } from "../../computation";
import { GLOBAL_CONFIG } from "../../config";
import { DragDropContext, Draggable, Droppable } from '@kanaries/react-beautiful-dnd';
import { COUNT_FIELD_ID, MEA_KEY_ID, MEA_VAL_ID, PAINT_FIELD_ID } from "../../constants";
import { getMeaAggName } from "../../utils";
const DEFAULT_SORT_ORDER = 'ascending';
function asDisplayValue(value) {
    if (value === null)
        return 'null';
    if (value === undefined)
        return '';
    if (typeof value === 'boolean')
        return value ? 'true' : 'false';
    return String(value);
}
const FieldConfigDialog = observer(() => {
    const vizStore = useVizStore();
    const computation = useCompututaion();
    const { t } = useTranslation();
    const target = vizStore.fieldConfigTarget;
    const isOpen = vizStore.showFieldConfigPanel && Boolean(target);
    const currentField = useMemo(() => {
        if (!target)
            return null;
        return vizStore.currentEncodings[target.channel]?.[target.index] ?? null;
    }, [target, vizStore.currentEncodings]);
    const [titleOverride, setTitleOverride] = useState('');
    const [sortType, setSortType] = useState('measure');
    const [sortOrder, setSortOrder] = useState('none');
    const [manualValues, setManualValues] = useState([]);
    const [loadingValues, setLoadingValues] = useState(false);
    const [fetchError, setFetchError] = useState(null);
    const [aggValue, setAggValue] = useState('sum');
    const [semanticTypeState, setSemanticTypeState] = useState('nominal');
    const [analyticTypeState, setAnalyticTypeState] = useState('dimension');
    const [customFormat, setCustomFormat] = useState('');
    const [windowAgg, setWindowAgg] = useState('none');
    const [manualSortDisabled, setManualSortDisabled] = useState(false);
    const [manualSortDisabledReason, setManualSortDisabledReason] = useState('');
    const [manualValuesFetched, setManualValuesFetched] = useState(false);
    const manualSortRequestIdRef = useRef(0);
    const [ready, setReady] = useState(false);
    const isMeasureField = analyticTypeState === 'measure';
    const isRowCountField = currentField?.fid === COUNT_FIELD_ID;
    const isInnerField = [MEA_VAL_ID, MEA_KEY_ID, PAINT_FIELD_ID].includes(currentField?.fid || '');
    const restrictToTitleAndFormatOnly = Boolean(isRowCountField);
    const disableTypeSelectors = Boolean(isInnerField || restrictToTitleAndFormatOnly);
    const isFetchingManualValues = loadingValues && sortType === 'manual';
    const defaultAggregated = vizStore.config.defaultAggregated;
    const allowWindowAgg = defaultAggregated &&
        isMeasureField &&
        semanticTypeState === 'quantitative' &&
        !isRowCountField &&
        !isInnerField;
    const allowSort = useMemo(() => {
        if (!target || !currentField)
            return false;
        if (analyticTypeState !== 'dimension')
            return false;
        if (semanticTypeState !== 'nominal' && semanticTypeState !== 'ordinal')
            return false;
        return target.channel === 'rows' || target.channel === 'columns';
    }, [target, analyticTypeState, semanticTypeState]);
    const specialManualValues = useMemo(() => {
        if (!currentField)
            return null;
        if (currentField.fid === MEA_KEY_ID) {
            const folds = vizStore.config.folds ?? [];
            const defaultAggregated = vizStore.config.defaultAggregated;
            const meaValField = vizStore.viewMeasures.find((field) => field.fid === MEA_VAL_ID);
            if (!folds.length)
                return [];
            return folds
                .map((fid) => vizStore.allFields.find((field) => field.fid === fid))
                .filter((field) => Boolean(field))
                .filter((field) => defaultAggregated || field.aggName !== 'expr')
                .map((field) => {
                if (!defaultAggregated) {
                    return {
                        value: field.name,
                        id: `${MEA_KEY_ID}_${field.fid}`,
                    };
                }
                const aggName = meaValField?.aggName ?? field.aggName;
                return {
                    value: getMeaAggName(field.name, aggName, field.windowAgg),
                    id: `${MEA_KEY_ID}_${field.fid}`,
                };
            });
        }
        if (currentField.fid === PAINT_FIELD_ID) {
            const mapParam = currentField.expression?.params?.find((param) => param.type === 'map' || param.type === 'newmap');
            if (!mapParam)
                return [];
            const paintMap = mapParam.value;
            const orderSource = paintMap.usedColor && paintMap.usedColor.length ? paintMap.usedColor : Object.keys(paintMap.dict).map((key) => Number(key));
            const colorOrder = orderSource.filter((key, index, arr) => !Number.isNaN(key) && arr.indexOf(key) === index);
            return colorOrder
                .map((key) => {
                const name = paintMap.dict[key]?.name;
                if (!name)
                    return null;
                return {
                    value: name,
                    id: `${PAINT_FIELD_ID}_${key}`,
                };
            })
                .filter((item) => Boolean(item));
        }
        return null;
    }, [currentField?.expression, currentField?.fid, vizStore.allFields, vizStore.config.defaultAggregated, vizStore.config.folds, vizStore.viewMeasures]);
    const aggregatorOptions = useMemo(() => GLOBAL_CONFIG.AGGREGATOR_LIST, []);
    const semanticTypeOptions = ['nominal', 'ordinal', 'quantitative', 'temporal'];
    const analyticTypeOptions = ['dimension', 'measure'];
    const windowAggOptions = [
        { value: 'none', label: 'None' },
        { value: 'running_total', label: 'Running total' },
        { value: 'difference', label: 'Difference' },
        { value: 'moving_average', label: 'Moving average' },
        { value: 'growth_rate', label: 'Compound growth rate' },
        { value: 'rank', label: 'Rank' },
    ];
    useEffect(() => {
        manualSortRequestIdRef.current += 1;
        if (!currentField) {
            setTitleOverride('');
            setSortType('measure');
            setSortOrder('none');
            setManualValues([]);
            setFetchError(null);
            setAggValue('sum');
            setSemanticTypeState('nominal');
            setAnalyticTypeState('dimension');
            setCustomFormat('');
            setManualSortDisabled(false);
            setManualSortDisabledReason('');
            setManualValuesFetched(false);
            setReady(false);
            return;
        }
        setTitleOverride(currentField.titleOverride ?? '');
        setSortType(currentField.sortType ?? 'measure');
        setSortOrder(currentField.sort ?? 'none');
        if (currentField.fid === MEA_KEY_ID || currentField.fid === PAINT_FIELD_ID) {
            setManualValues([]);
        }
        else {
            setManualValues((currentField.sortList ?? []).map((value, idx) => ({ value, id: `${currentField.fid}_${idx}` })));
        }
        setFetchError(null);
        setAggValue(currentField.aggName ?? 'sum');
        setSemanticTypeState(currentField.semanticType);
        setAnalyticTypeState(currentField.analyticType);
        setCustomFormat(currentField.customFormat ?? '');
        setWindowAgg(currentField.windowAgg ?? 'none');
        setManualSortDisabled(false);
        setManualSortDisabledReason('');
        setManualValuesFetched(false);
        setReady(true);
    }, [currentField?.fid, target?.channel, target?.index]);
    const handleAnalyticTypeChange = useCallback((value) => {
        setAnalyticTypeState(value);
        if (value === 'dimension') {
            setAggValue('sum');
            setWindowAgg('none');
        }
        else if (value === 'measure') {
            setAggValue(GLOBAL_CONFIG.AGGREGATOR_LIST[0] ?? 'sum');
        }
    }, [aggValue]);
    useEffect(() => {
        if (!allowWindowAgg && windowAgg !== 'none') {
            setWindowAgg('none');
        }
    }, [allowWindowAgg, windowAgg]);
    const hydrateValues = useCallback(async () => {
        if (!currentField || !allowSort || !ready || sortType !== 'manual')
            return;
        setFetchError(null);
        if (specialManualValues !== null) {
            setManualValuesFetched(true);
            setLoadingValues(false);
            const distinctTotal = specialManualValues.length;
            const limitExceeded = distinctTotal > 100;
            setManualSortDisabled(limitExceeded);
            setManualSortDisabledReason(limitExceeded ? `Manual sort supports up to 100 distinct values. ${distinctTotal} values detected.` : '');
            if (limitExceeded) {
                setManualValues([]);
                setSortType((prev) => (prev === 'manual' ? 'measure' : prev));
                return;
            }
            setManualValues((prev) => {
                if (prev.length > 0)
                    return prev;
                return specialManualValues;
            });
            return;
        }
        const requestId = ++manualSortRequestIdRef.current;
        setManualValuesFetched(true);
        setLoadingValues(true);
        try {
            const stats = await fieldStat(computation, currentField, {
                values: true,
                range: false,
                valuesMeta: true,
                valuesLimit: 101,
            }, vizStore.meta);
            if (requestId !== manualSortRequestIdRef.current) {
                return;
            }
            const distinctTotal = stats.valuesMeta?.distinctTotal ?? stats.values.length;
            const limitExceeded = distinctTotal > 100;
            setManualSortDisabled(limitExceeded);
            setManualSortDisabledReason(limitExceeded ? `Manual sort supports up to 100 distinct values. ${distinctTotal} values detected.` : '');
            if (limitExceeded) {
                setManualValues([]);
                setSortType((prev) => (prev === 'manual' ? 'measure' : prev));
                return;
            }
            if (stats.values.length) {
                setManualValues((prev) => {
                    if (prev.length > 0)
                        return prev;
                    return stats.values.map((v, idx) => ({
                        value: v.value,
                        id: `${currentField.fid}_auto_${idx}`,
                    }));
                });
            }
        }
        catch (err) {
            if (requestId === manualSortRequestIdRef.current) {
                setFetchError(err instanceof Error ? err.message : String(err));
            }
        }
        finally {
            if (requestId === manualSortRequestIdRef.current) {
                setLoadingValues(false);
            }
        }
    }, [allowSort, ready, computation, currentField, sortType, vizStore.meta, specialManualValues]);
    useEffect(() => {
        if (isOpen && allowSort && currentField && sortType === 'manual' && !manualValuesFetched && !loadingValues) {
            hydrateValues();
        }
    }, [allowSort, currentField?.fid, hydrateValues, isOpen, loadingValues, manualValuesFetched, sortType, target?.channel, target?.index]);
    useEffect(() => {
        if (!currentField || sortType !== 'manual' || specialManualValues === null)
            return;
        const distinctTotal = specialManualValues.length;
        const limitExceeded = distinctTotal > 100;
        setManualSortDisabled(limitExceeded);
        setManualSortDisabledReason(limitExceeded ? `Manual sort supports up to 100 distinct values. ${distinctTotal} values detected.` : '');
        setManualValuesFetched(true);
        setLoadingValues(false);
        setFetchError(null);
        if (limitExceeded) {
            setManualValues([]);
            setSortType((prev) => (prev === 'manual' ? 'measure' : prev));
            return;
        }
        setManualValues((prev) => {
            if (prev.length === 0) {
                return specialManualValues;
            }
            const isSame = prev.length === specialManualValues.length &&
                prev.every((item, index) => item.id === specialManualValues[index].id && item.value === specialManualValues[index].value);
            if (isSame)
                return prev;
            const specialMap = new Map(specialManualValues.map((item) => [item.id, item]));
            const ordered = [];
            prev.forEach((item) => {
                const replacement = specialMap.get(item.id);
                if (replacement) {
                    ordered.push(replacement);
                    specialMap.delete(item.id);
                }
            });
            specialMap.forEach((item) => ordered.push(item));
            return ordered;
        });
    }, [currentField, sortType, specialManualValues]);
    const handleRetryManualValues = useCallback(() => {
        if (loadingValues)
            return;
        setFetchError(null);
        setManualValuesFetched(false);
    }, [loadingValues]);
    const handleClose = useCallback(() => {
        vizStore.setShowFieldConfigPanel(false);
    }, [vizStore]);
    const handleSave = useCallback(() => {
        if (!target || !currentField)
            return;
        const trimmedTitle = titleOverride.trim();
        const trimmedFormat = customFormat.trim();
        const patch = {
            titleOverride: trimmedTitle ? trimmedTitle : undefined,
            semanticType: semanticTypeState,
            analyticType: analyticTypeState,
            customFormat: trimmedFormat ? trimmedFormat : undefined,
        };
        if (analyticTypeState === 'measure') {
            patch.aggName = aggValue;
        }
        else if (currentField.aggName) {
            patch.aggName = undefined;
        }
        if (allowSort) {
            if (sortType === 'manual') {
                patch.sortType = 'manual';
                patch.sortList = manualValues.map((item) => item.value);
                patch.sort = patch.sortList.length ? DEFAULT_SORT_ORDER : 'none';
            }
            else if (sortType === 'alphabetical') {
                patch.sortType = 'alphabetical';
                patch.sort = sortOrder === 'none' ? DEFAULT_SORT_ORDER : sortOrder;
                patch.sortList = undefined;
            }
            else {
                patch.sortType = 'measure';
                patch.sort = sortOrder;
                patch.sortList = undefined;
            }
        }
        if (allowWindowAgg && windowAgg !== 'none') {
            patch.windowAgg = windowAgg;
        }
        else if (currentField.windowAgg) {
            patch.windowAgg = undefined;
        }
        vizStore.editEncodingField(target.channel, target.index, patch);
        handleClose();
    }, [
        allowSort,
        allowWindowAgg,
        currentField,
        handleClose,
        manualValues,
        semanticTypeState,
        analyticTypeState,
        aggValue,
        sortOrder,
        sortType,
        target,
        titleOverride,
        customFormat,
        windowAgg,
        vizStore,
    ]);
    const reorderManualValue = useCallback((from, to) => {
        setManualValues((prev) => {
            if (from === to)
                return prev;
            const next = prev.slice();
            const [item] = next.splice(from, 1);
            next.splice(to, 0, item);
            return next;
        });
    }, []);
    const renderSortOrderControls = () => (React.createElement("div", { className: "space-y-2" },
        React.createElement(Label, null, "Sort Order"),
        React.createElement(RadioGroup, { value: sortOrder, onValueChange: (value) => setSortOrder(value), className: "flex space-x-3" },
            React.createElement("div", { className: "flex items-center space-x-2" },
                React.createElement(RadioGroupItem, { value: "ascending", id: "order-asc" }),
                React.createElement(Label, { htmlFor: "order-asc" }, "Ascending")),
            React.createElement("div", { className: "flex items-center space-x-2" },
                React.createElement(RadioGroupItem, { value: "descending", id: "order-desc" }),
                React.createElement(Label, { htmlFor: "order-desc" }, "Descending")),
            sortType === 'measure' && (React.createElement("div", { className: "flex items-center space-x-2" },
                React.createElement(RadioGroupItem, { value: "none", id: "order-none" }),
                React.createElement(Label, { htmlFor: "order-none" }, "None"))))));
    const handleManualDragEnd = useCallback((result) => {
        if (!result.destination)
            return;
        reorderManualValue(result.source.index, result.destination.index);
    }, [reorderManualValue]);
    const manualList = (React.createElement("div", { className: "space-y-3" },
        React.createElement("div", { className: "flex items-end space-x-2" },
            React.createElement("div", { className: "flex-1" },
                React.createElement(Label, { htmlFor: "manual-input" }, "Custom Value"))),
        React.createElement("div", { className: "rounded border p-2" },
            React.createElement("div", { className: "flex items-center justify-between mb-2 text-xs text-muted-foreground" },
                React.createElement("span", null, "Drag to change order")),
            fetchError && (React.createElement("div", { className: "flex items-center justify-between mb-2 text-xs text-destructive" },
                React.createElement("span", { className: "pr-2" }, fetchError),
                React.createElement("button", { type: "button", className: "text-primary underline", onClick: handleRetryManualValues }, "Retry"))),
            React.createElement(ScrollArea, { className: "max-h-48" },
                React.createElement(DragDropContext, { onDragEnd: handleManualDragEnd },
                    React.createElement(Droppable, { droppableId: "manual-values", direction: "vertical" }, (provided) => (React.createElement("div", { ref: provided.innerRef, ...provided.droppableProps, className: "space-y-1" },
                        manualValues.map((item, index) => (React.createElement(Draggable, { key: item.id, draggableId: item.id, index: index }, (dragProvided, snapshot) => (React.createElement("div", { ref: dragProvided.innerRef, className: `flex items-center justify-between rounded border px-2 py-1 text-sm transition-colors ${snapshot.isDragging ? 'bg-primary/10 border-primary' : 'bg-muted/40'}`, ...dragProvided.draggableProps, ...dragProvided.dragHandleProps },
                            React.createElement("span", { className: "truncate pr-4" }, asDisplayValue(item.value))))))),
                        provided.placeholder,
                        isFetchingManualValues && (React.createElement("span", { className: "flex items-center space-x-1 text-primary" },
                            React.createElement(Spinner, { className: "h-3 w-3" }),
                            React.createElement("span", null, "Loading\u2026")))))))))));
    const sortControls = allowSort ? (React.createElement("div", { className: "space-y-4" },
        React.createElement("div", null,
            React.createElement(Label, null, "Custom Sort"),
            React.createElement(RadioGroup, { value: sortType, onValueChange: (value) => setSortType(value), className: "space-y-2 mt-2" },
                React.createElement("div", { className: "flex items-center space-x-2" },
                    React.createElement(RadioGroupItem, { value: "measure", id: "sort-measure" }),
                    React.createElement(Label, { htmlFor: "sort-measure" }, "By opposite axis (default)")),
                React.createElement("div", { className: "flex items-center space-x-2" },
                    React.createElement(RadioGroupItem, { value: "alphabetical", id: "sort-alpha" }),
                    React.createElement(Label, { htmlFor: "sort-alpha" }, "Alphabetical")),
                manualSortDisabled ? (React.createElement(Tooltip, { content: manualSortDisabledReason || 'Manual sort supports up to 100 distinct values.' },
                    React.createElement("div", { className: "flex items-center space-x-2 text-muted-foreground" },
                        React.createElement(RadioGroupItem, { value: "manual", id: "sort-manual", disabled: true }),
                        React.createElement(Label, { htmlFor: "sort-manual" }, "Manual order")))) : (React.createElement("div", { className: "flex items-center space-x-2" },
                    React.createElement(RadioGroupItem, { value: "manual", id: "sort-manual" }),
                    React.createElement(Label, { htmlFor: "sort-manual" }, "Manual order"),
                    isFetchingManualValues && React.createElement(Spinner, { className: "h-3 w-3 text-primary" }))))),
        (sortType === 'measure' || sortType === 'alphabetical') && renderSortOrderControls(),
        sortType === 'manual' && manualList)) : null;
    return (React.createElement(Dialog, { open: isOpen, onOpenChange: (open) => (!open ? handleClose() : undefined) },
        React.createElement(DialogContent, { className: "max-w-xl" },
            React.createElement(DialogHeader, null,
                React.createElement(DialogTitle, null, "Field Settings"),
                currentField && React.createElement("p", { className: "text-sm text-muted-foreground" }, currentField.name)),
            currentField ? (React.createElement("div", { className: "space-y-6 my-2" },
                React.createElement("div", { className: "space-y-2" },
                    React.createElement(Label, { htmlFor: "field-title" }, "Title Override"),
                    React.createElement(Textarea, { id: "field-title", value: titleOverride, onChange: (event) => setTitleOverride(event.target.value), placeholder: "Leave empty to use the default title" })),
                React.createElement("div", { className: "grid gap-4 sm:grid-cols-2" },
                    React.createElement("div", { className: "space-y-2" },
                        React.createElement(Label, null, "Analytic Type"),
                        React.createElement(Select, { value: analyticTypeState, onValueChange: (value) => handleAnalyticTypeChange(value), disabled: disableTypeSelectors },
                            React.createElement(SelectTrigger, null,
                                React.createElement(SelectValue, { placeholder: "Select analytic type" })),
                            React.createElement(SelectContent, null, analyticTypeOptions.map((option) => (React.createElement(SelectItem, { key: option, value: option }, option)))))),
                    React.createElement("div", { className: "space-y-2" },
                        React.createElement(Label, null, "Semantic Type"),
                        React.createElement(Select, { value: semanticTypeState, onValueChange: (value) => setSemanticTypeState(value), disabled: disableTypeSelectors },
                            React.createElement(SelectTrigger, null,
                                React.createElement(SelectValue, { placeholder: "Select semantic type" })),
                            React.createElement(SelectContent, null, semanticTypeOptions.map((option) => (React.createElement(SelectItem, { key: option, value: option }, option)))))),
                    React.createElement("div", { className: "space-y-2" },
                        React.createElement(Label, null, "Aggregator"),
                        React.createElement(Select, { value: aggValue, onValueChange: (value) => setAggValue(value), disabled: (!isMeasureField && !isInnerField) || isRowCountField },
                            React.createElement(SelectTrigger, null,
                                React.createElement(SelectValue, { placeholder: "Select aggregator" })),
                            React.createElement(SelectContent, null, aggregatorOptions.map((option) => (React.createElement(SelectItem, { key: option, value: option }, t(`constant.aggregator.${option}`)))))),
                        isRowCountField ? (React.createElement("p", { className: "text-xs text-muted-foreground" }, "Row count uses a fixed aggregator.")) : (!isMeasureField && React.createElement("p", { className: "text-xs text-muted-foreground" }, "Aggregator only applies to measure fields."))),
                    defaultAggregated && (React.createElement("div", { className: "space-y-2" },
                        React.createElement(Label, null, "Window Aggregation"),
                        React.createElement(Select, { value: windowAgg, onValueChange: (value) => setWindowAgg(value), disabled: !allowWindowAgg },
                            React.createElement(SelectTrigger, null,
                                React.createElement(SelectValue, { placeholder: "Select window aggregation" })),
                            React.createElement(SelectContent, null, windowAggOptions.map((option) => (React.createElement(SelectItem, { key: option.value, value: option.value }, option.label))))))),
                    React.createElement("div", { className: "space-y-2" },
                        React.createElement(Label, { htmlFor: "field-format" }, "Custom Format"),
                        React.createElement(Input, { id: "field-format", value: customFormat, placeholder: "e.g. ,.2f or %Y-%m-%d", onChange: (event) => setCustomFormat(event.target.value) }),
                        React.createElement("p", { className: "text-xs text-muted-foreground" },
                            t(`config.formatGuidesDocs`),
                            ":",
                            ' ',
                            React.createElement("a", { target: "_blank", className: "hover:underline text-primary", href: "https://github.com/d3/d3-format#locale_format" }, t(`config.readHere`))))),
                sortControls)) : (React.createElement("p", { className: "text-sm text-muted-foreground" }, "No field selected.")),
            React.createElement(DialogFooter, null,
                React.createElement(Button, { variant: "secondary", onClick: handleClose }, t('actions.cancel')),
                React.createElement(Button, { onClick: handleSave, disabled: !currentField }, t('actions.confirm'))))));
});
export default FieldConfigDialog;
